shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt

```diff
@@ -0,0 +1,63 @@
+package com.ivy.domain.usecase
+
+import com.ivy.data.model.primitive.AssetCode
+import com.ivy.data.model.primitive.NonZeroDouble
+import com.ivy.data.model.primitive.PositiveDouble
+
+class BalanceBuilder {
+
+    private val balance = mutableMapOf<AssetCode, NonZeroDouble>()
+
+    fun processDeposit(
+        incomes: Map<AssetCode, PositiveDouble>,
+        transferIn: Map<AssetCode, PositiveDouble>,
+    ) {
+        getAmount(incomes, transferIn).forEach { (asset, value) ->
+            NonZeroDouble
+                .from((balance[asset]?.value ?: 0.0) + value.value)
+                .onRight { newValue ->
+                    balance[asset] = newValue
+                }
+        }
+    }
+
+    fun processWithdrawal(
+        expenses: Map<AssetCode, PositiveDouble>,
+        transferOut: Map<AssetCode, PositiveDouble>,
+    ) {
+        getAmount(expenses, transferOut).forEach { (asset, value) ->
+            NonZeroDouble
+                .from((balance[asset]?.value ?: 0.0) + (-value.value))
+                .onRight { newValue ->
+                    balance[asset] = newValue
+                }
+        }
+    }
+
+    private fun getAmount(
```

Comment:
That's `combine` tbh. Just copy-paste the contents of `combine` here and we can end-up with just one function. Also `a` and `b` IMO are better names for the maps. From math POV:
- combine: a + b -> c

Combines two arbitrary maps into a third one

Reply:
Fair enough!